### ISSUE TITLE
[VoxelGridFilter] Use different LeafSize for different axes

### ISSUE THREAD

**Issue by mszuyx** [PointCloudLibrary/pcl#5581]
**Describe the bug**

I am trying to use a non-cubical grid when downsampling a pointcloud. (i.e., use different leaf size setting for x, y, and z axis)
For example:
  ```
pcl::VoxelGrid<pcl::PointXYZ> sor;
  sor.setInputCloud (cloud);
  sor.setLeafSize (float (0.05),float (1),float (0.05));
  sor.filter (*cloud);
```
Ideally I will get a filtered pointcloud with less density in y axis as I used a "long" grid in the corresponding axis.
But I can tell any difference between the "cubic" grid and the "long" grid settings (the pointcloud density is always uniform in all axes despite the following settings):
` sor.setLeafSize (float (0.05),float (0.05),float (0.05));`
VS
` sor.setLeafSize (float (0.05),float (1),float (0.05));`

I wonder if there are some hidden rules inside VoxelGrid that automatically corrects the Leaf size setting as it always "expect" the input to be cubical. Or, maybe my assumption about "larger leaf size = lower density in the corresponding axis" is wrong.

If my assumption is wrong, is there another function I can use to reduce pointcloud density/resolution in a specific axis?

Thanks!,
Yu

**Comment by mvieth** [comment]
@mszuyx What do you mean with "lower density in a specific axis"? Can you maybe post pictures of the output clouds with both settings? How many points remain after filtering in each setting?

**Comment by mszuyx** [comment]
@mvieth I took some screenshots:

Here is the original scene before the voxel filter:
![original](https://user-images.githubusercontent.com/37651144/213052967-72f89d79-efcc-40c4-a3cd-b280211c6e91.png)

Here is the point cloud after the voxel filter with the following settings (non-cubical grid):
` setLeafSize (float (0.03),float (2.0),float (0.03));`
The point cloud size is 3060
![non_cube](https://user-images.githubusercontent.com/37651144/213053178-3a045502-d4cd-4cff-ba4d-7e6036cfc742.png)


Here is the point cloud after the voxel filter with the following settings (cubical grid):
` setLeafSize (float (0.03),float (0.03),float (0.03));`
The point cloud size is 3549
![cube](https://user-images.githubusercontent.com/37651144/213053184-0d32ce36-8987-45c8-b8cb-6052185d1a92.png)

As expected, a larger leaf size will result in fewer point cloud size as the downsampling is more "aggressive". 
However, with an extreme (2m) leaf size in one axis, I think it is reasonable to expect some "slicing" effects and a much smaller point cloud size. Yet, the difference in point cloud size is minor, and the resulting point cloud still looks "homogeneous" in all axes.

To clarify, the "slicing" effect is what I am looking for in this application (desired behaviour).

**Comment by mvieth** [comment]
@mszuyx  For completeness: how many points are in your original scene? Can you also show the axes so I can see which direction the y-axis is? Have you also tried a larger leaf size in the x- and z-axis (while setting the other two axes to 0.03, respectively)?

I believe downsampling with a larger leaf size in one dimension works, but it might be difficult to see the effects in your point cloud. E.g., if you had a plane (table) in the x-y plane, and you would downsample with leaf sizes 0.03, 2.0, 0.03, it would probably be easy to see the effects.
And I am still not sure what you mean with "slicing effect", or whether the voxel grid filter is the way to achieve it.

**Comment by mszuyx** [comment]
@mvieth Sorry, my bad.

I am running the experiment off an RGB-D camera, so the point cloud size is inconsistent each time I run it.
For this scene, it is roughly 6,000 points.
![Screenshot 2023-01-18 132106](https://user-images.githubusercontent.com/37651144/213275121-58c2ab18-f8df-4056-a276-69ab6324ff73.png)

The y-axis is the vertical axis (green pointing downward) here. The z-axis is the depth direction, and the x-axis is the "left-right" direction. 

> Have you also tried a larger leaf size in the x- and z-axis (while setting the other two axes to 0.03, respectively)?
For that,

Using the setting:
`setLeafSize (float (2),float (0.03),float (0.03));`
The point cloud size is 2270.
![Screenshot 2023-01-18 133804](https://user-images.githubusercontent.com/37651144/213280199-ffc71e58-d3ab-4e89-843e-4a85257f9ba8.png)

Using the setting:
`setLeafSize (float (0.03),float (0.03),float (2));`
The point cloud size is 2506.
![Screenshot 2023-01-18 134440](https://user-images.githubusercontent.com/37651144/213280215-dc38018f-fc8b-4870-a220-e31b5c3d85a4.png)

Using the setting:
`setLeafSize (float (2),float (2),float (2));`
The point cloud size is 26.
![Screenshot 2023-01-18 134701](https://user-images.githubusercontent.com/37651144/213280476-ca286295-a32b-4b33-aacd-cf8b0d27e3b8.png)


>  if you had a plane (table) in the x-y plane, and you would downsample with leaf sizes 0.03, 2.0, 0.03, it would probably be easy to see the effects. 

Agree. The idea here is if we have horizontal planes (e.g., tabletops) and vertical planes (e.g., walls), more aggressive downsampling in the vertical axis will result in:
1. The horizontal planes are mostly unaffected (still look like planes) if they are perpendicular to the vertical axis (in my case, the y-axis).
2. The vertical planes will be broken down and turned into horizontal line segments (That's what I meant by "slicing").
At the moment, I don't have an easy way to set up a good visualization. I hope this description helps.
The bottom line is, with an extreme setting like this (0.03, 2.0, 0.03), people should be able to tell each axis got downsampled the most quite easily. Yet from my screenshots, it doesn't seem so.

> whether the voxel grid filter is the way to achieve it.

In my understanding, the voxel grids used in the voxel grid filter do not overlap. Then I believe a non-cubical gird size setting can achieve this effect. If not, do you have any suggestions？

**Comment by mszuyx** [comment]
> There is no guarantee that the voxel grid output of vertical planes will form a straight line. Remember that the voxel grid will approximate with the centroid/average of all points in each voxel. So at best you will get a very noisy/wobbly line for vertical planes.

Yes, I do understand. The original statement is more like a figure of speech.

> Judging from the name, you are using the frame of your depth sensor. Are you absolutely sure that this frame is correct, or do you need a transformation to arrive at a "world frame"? That is, are you sure the y-axis is perpendicular to the ground? If it isn't, you will likely not be able to achieve the result you want.

The point cloud I am working with has been through an IMU orientation alignment such that the vertical axis is perpendicular to the ground. And by using a box crop filter, I verified that the vertical axis is the y-axis.

Here are some more figures for your reference:
The filtered point cloud used the setting:
` setLeafSize (float (0.03),float (2.0),float (0.03));`
![Picture1](https://user-images.githubusercontent.com/37651144/213551118-fe043155-766c-421c-971c-a38242104923.png)

From these images, I think we can learn two things:

1. The ground plane in this point cloud is mostly perpendicular to the vertical axis.
2. To some degree, the voxel filter with the setting ` setLeafSize (float (0.03),float (2.0),float (0.03));` did filter out some vertical features and left the ground plane mostly unaffected.

BUT, again, with a leaf size setting this extreme, I expect a more dramatic change in point cloud density in the vertical axis.

> Larger leaf size in y-direction (ground plane unaffected, but the objects have fever points):

That's exactly what I am looking for! I am using the voxel grid filter as a pre-processing step to help isolate the ground plane.
Is it possible for you to share a side or "isometric" view of the same scene? And also your leaf size setting?

> So the voxel grid with different leaf sizes works, and it could be used to achieve what you want (if I understood correctly what you want). Except that vertical planes wont be perfectly straight lines.

Agree. For my application, I only care about the ground plane. As long as we can find an effective way to break down the vertical features (straight lines or not), it's fine!

**Comment by mvieth** [comment]
> Is it possible for you to share a side or "isometric" view of the same scene? And also your leaf size setting?

These are the commands I used:
```bash
pcl_transform_point_cloud milk_cartoon_all_small_clorox.pcd transformed.pcd -axisangle 1,0,0,-0.608
pcl_voxel_grid transformed.pcd filtered1.pcd -leaf 1.0,0.01,0.01
pcl_voxel_grid transformed.pcd filtered2.pcd -leaf 0.01,1.0,0.01
pcl_voxel_grid transformed.pcd filtered3.pcd -leaf 0.01,0.01,1.0
```

**Comment by mszuyx** [comment]
@mvieth thanks!
